app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/recordaudio/AudioMediaRecorder.kt

```diff
@@ -209,13 +222,100 @@ class AudioMediaRecorder @Inject constructor(
+        mediaExtractor.setDataSource(inputFilePath)
+
+        val format = MediaFormat.createAudioFormat(
+            MediaFormat.MIMETYPE_AUDIO_AAC,
+            SAMPLING_RATE,
+            AUDIO_CHANNELS
+        )
+        format.setInteger(MediaFormat.KEY_BIT_RATE, BIT_RATE)
+        format.setInteger(MediaFormat.KEY_AAC_PROFILE, MediaCodecInfo.CodecProfileLevel.AACObjectLC)
+
+        val codec = MediaCodec.createEncoderByType(MediaFormat.MIMETYPE_AUDIO_AAC)
+        codec.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
+        codec.start()
+
+        val bufferInfo = MediaCodec.BufferInfo()
+        val muxer = MediaMuxer(mp4OutputPath.toString(), MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
+        var trackIndex = -1
+        var sawInputEOS = false
+        var sawOutputEOS = false
+
+        val fileInputStream = FileInputStream(inputFile)
+
+        while (!sawOutputEOS) {
+            if (!sawInputEOS) {
+                val inputBufferIndex = codec.dequeueInputBuffer(TIMEOUT_US)
+                if (inputBufferIndex >= 0) {
+                    val inputBuffer = codec.getInputBuffer(inputBufferIndex)
+                    inputBuffer?.clear()
+
+                    val sampleSize = fileInputStream.channel.read(inputBuffer!!)
+                    if (sampleSize < 0) {
+                        codec.queueInputBuffer(inputBufferIndex, 0, 0, 0, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
+                        sawInputEOS = true
+                    } else {
+                        codec.queueInputBuffer(inputBufferIndex, 0, sampleSize, System.nanoTime() / 1000, 0)
+                    }
+                }
+            }
+
+            val outputBufferIndex = codec.dequeueOutputBuffer(bufferInfo, TIMEOUT_US)
+            if (outputBufferIndex >= 0) {
+                val outputBuffer = codec.getOutputBuffer(outputBufferIndex)
+
+                if (bufferInfo.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG != 0) {
+                    codec.releaseOutputBuffer(outputBufferIndex, false)
+                    continue
+                }
+
+                if (bufferInfo.size != 0) {
+                    outputBuffer?.position(bufferInfo.offset)
+                    outputBuffer?.limit(bufferInfo.offset + bufferInfo.size)
+
+                    if (trackIndex == -1) {
+                        val newFormat = codec.outputFormat
+                        trackIndex = muxer.addTrack(newFormat)
+                        muxer.start()
+                    }
+
+                    muxer.writeSampleData(trackIndex, outputBuffer!!, bufferInfo)
+                }
+
+                codec.releaseOutputBuffer(outputBufferIndex, false)
+
+                if (bufferInfo.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) {
+                    sawOutputEOS = true
+                }
+            }
+        }
+
+        fileInputStream.close()
+        muxer.stop()
+        muxer.release()
+        codec.stop()
+        codec.release()
```

Comment:
Question: Can we wrap this function into a big `try-catch` ? (or preferred, alternative) would do it. 
My concern is in case of failure we leave the resources open, leading to leaks.

Reply:
I added big try catch and left logger to log those failures. If it will fail then messages will be send with wav format